clients/application/src/main/java/org/eclipse/hono/client/application/DownstreamMessage.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) 2021 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+
+package org.eclipse.hono.client.application;
+
+import java.util.Objects;
+
+import io.vertx.core.buffer.Buffer;
+
+/**
+ * A message of Hono's northbound APIs, flowing from the messaging system to the backend application.
+ */
+public abstract class DownstreamMessage implements Message {
+
+    private final String tenantId;
+    private final String deviceId;
+    private final MessageProperties properties;
+    private final Buffer payload;
+    private final String contentType;
+    private final MessageContext messageContext;
+
+    /**
+     * Creates a downstream message.
+     *
+     * @param tenantId The tenant that sent the message.
+     * @param deviceId The device that sent the message.
+     * @param properties The metadata of the message.
+     * @param payload The payload.
+     * @param contentType The content type of the message payload.
+     * @param messageContext The context to be added to the message.
+     * @throws NullPointerException if any of the parameters is {@code null}.
+     */
+    public DownstreamMessage(final String tenantId, final String deviceId, final MessageProperties properties,
+            final Buffer payload, final String contentType, final MessageContext messageContext) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(deviceId);
+        Objects.requireNonNull(properties);
+        Objects.requireNonNull(payload);
```

Comment:
IMHO we support empty messages (without a payload) as well, e.g. Empty Notifications and empty command response messages